File: app/schemas/institutional_signals.py
"""机构资金信号（Institutional Signals）Pydantic schemas。

产品只对外暴露「状态 + 五维评分 + 证据链」，原始数据折叠在 signals 明细里。
"""
from typing import List, Optional

from pydantic import Field

from app.schemas.base import BaseResponse

# 五个维度的稳定 key，与 dimensions.py 一一对应
DIMENSION_KEYS = ("expectation", "positioning", "participation", "fundamental", "confirmation")


class SignalItem(BaseResponse):
    """单条底层信号（如 Relative Volume、Target Price）。"""

    key: str = Field(description="稳定标识，如 relative_volume / target_price")
    label: str = Field(description="中文名，如 相对成交量")
    value: Optional[str] = Field(None, description="展示值，如 2.1x / +3.4%")
    direction: str = Field("flat", description="方向：up / down / flat")
    hit: bool = Field(False, description="是否触发（用于状态组合判定）")
    detail: Optional[str] = Field(None, description="补充说明")


class DimensionScore(BaseResponse):
    """五维之一的评分与明细。"""

    key: str = Field(description="维度 key，见 DIMENSION_KEYS")
    label: str = Field(description="维度中文名，如 预期")
    question: str = Field(description="该维度回答的核心问题")
    score: float = Field(ge=0, le=100, description="子分 0–100")
    status: str = Field("ok", description="ok / partial / unavailable")
    signals: List[SignalItem] = Field(default_factory=list)


class SignalState(BaseResponse):
    """由维度组合推导出的状态标签（如 机构建仓 / 趋势确认）。"""

    key: str = Field(description="状态 key，如 expectation_upgrade")
    emoji: str
    label: str = Field(description="状态中文名")
    stars: int = Field(ge=1, le=5, description="重要度星级")
    meaning: str = Field(description="一句话含义")
    evidence: List[str] = Field(default_factory=list, description="命中的证据链")


class InstitutionalSignalReport(BaseResponse):
    """GET /api/v1/institutional-signals?symbol= 完整响应。"""

    symbol: str
    name: str
    as_of: str = Field(description="报告日期 YYYY-MM-DD")
    composite_score: float = Field(ge=0, le=100, description="五维加权综合分")
    headline: str = Field(description="一句话结论")
    dimensions: List[DimensionScore]
    states: List[SignalState]

File: frontend/lib/api/institutional_signals.ts
import apiClient from './client'

export interface SignalExplanation {
  inputs: string[]
  formula: string | null
  thresholds: string | null
  conclusion: string | null
  source: string | null
}

export interface SignalItem {
  key: string
  label: string
  value: string | null
  direction: 'up' | 'down' | 'flat'
  hit: boolean
  detail: string | null
  explain: SignalExplanation | null
}

export interface DimensionScore {
  key: string
  label: string
  question: string
  score: number
  status: 'ok' | 'partial' | 'unavailable'
  signals: SignalItem[]
}

export interface SignalState {
  key: string
  emoji: string
  label: string
  stars: number
  meaning: string
  logic: string | null
  evidence: string[]
  buy_rank: number | null
  buy_timing: string | null
  buy_edge: string | null
  buy_thesis: string | null
}

export interface BuyStage {
  key: string
  emoji: string
  label: string
  timing: string
  edge: string
  thesis: string
  rank: number
  active: boolean
}

export interface InstitutionalSignalReport {
  request_id: string
  symbol: string
  name: string
  as_of: string
  composite_score: number
  coverage: number
  coverage_total: number
  confidence: string
  headline: string
  buy_headline: string
  buy_ladder: BuyStage[]
  price_history: number[]
  dimensions: DimensionScore[]
  states: SignalState[]
}

export async function fetchInstitutionalSignals(
  symbol: string,
  refresh: boolean = false,
): Promise<InstitutionalSignalReport> {
  const { data } = await apiClient.get<InstitutionalSignalReport>(
    '/api/v1/institutional-signals',
    { params: { symbol, ...(refresh ? { refresh: true } : {}) }, timeout: 30000 },
  )
  return data
}

export interface LeaderboardEntry {
  symbol: string
  name: string
  composite_score: number
  coverage: number
  confidence: string
  top_state: SignalState | null
  states: SignalState[]
  dimension_scores: Record<string, number>
}

export interface LeaderboardResponse {
  request_id: string
  status: string
  as_of: string
  computed_at: string
  universe_source: string
  universe_size: number
  scanned: number
  note: string
  entries: LeaderboardEntry[]
}

export async function fetchLeaderboard(
  universe: string = 'sp500',
  refresh: boolean = false,
): Promise<LeaderboardResponse> {
  const { data } = await apiClient.get<LeaderboardResponse>(
    '/api/v1/institutional-signals/leaderboard',
    { params: { universe, ...(refresh ? { refresh: true } : {}) }, timeout: 90000 }, // 首次扫描 universe 较慢，给足 90s
  )
  return data
}
